ecdsa/signing/finalize.go

```diff
@@ -7,73 +7,323 @@
 package signing
 
 import (
+	"bytes"
 	"crypto/ecdsa"
 	"errors"
 	"fmt"
 	"math/big"
 
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/hashicorp/go-multierror"
+
 	"github.com/binance-chain/tss-lib/common"
+	"github.com/binance-chain/tss-lib/crypto"
+	"github.com/binance-chain/tss-lib/crypto/paillier"
+	"github.com/binance-chain/tss-lib/crypto/zkp"
 	"github.com/binance-chain/tss-lib/tss"
 )
 
-func (round *finalization) Start() *tss.Error {
-	if round.started {
-		return round.WrapError(errors.New("round already started"))
+const (
+	TaskNameFinalize = "signing-finalize"
+)
+
+// -----
+// One Round Finalization (async/offline)
+// -----
+
+// FinalizeGetOurSigShare is called in one-round signing mode after the online rounds have finished to compute s_i.
+func FinalizeGetOurSigShare(state *SignatureData, msg *big.Int) (sI *big.Int) {
+	data := state.GetOneRoundData()
+
+	N := tss.EC().Params().N
+	modN := common.ModInt(N)
+
+	kI, rSigmaI := new(big.Int).SetBytes(data.GetKI()), new(big.Int).SetBytes(data.GetRSigmaI())
+	sI = modN.Add(modN.Mul(msg, kI), rSigmaI)
+	return
+}
+
+// FinalizeGetOurSigShare is called in one-round signing mode to build a final signature given others' s_i shares and a msg.
+// Note: each P in otherPs should correspond with that P's s_i at the same index in otherSIs.
+func FinalizeGetAndVerifyFinalSig(
+	state *SignatureData,
+	pk *ecdsa.PublicKey,
+	msg *big.Int,
+	ourP *tss.PartyID,
+	ourSI *big.Int,
+	otherSIs map[*tss.PartyID]*big.Int,
+) (*SignatureData, *btcec.Signature, *tss.Error) {
+	if len(otherSIs) == 0 {
+		return nil, nil, FinalizeWrapError(errors.New("len(otherSIs) == 0"), ourP)
+	}
+	data := state.GetOneRoundData()
+	if data.GetT() != int32(len(otherSIs)) {
```

Comment:
in this lib we actually support any number of signers >= t+1 and <= n

Reply:
This insures when we do the online signing later, we have the same number of participants. it is not the threshold of the tss.